<div>
Olá, Vitor!

Meu nome é Luiz. Fico feliz em revisar seu projeto. Ao longo do texto farei algumas observações sobre melhorias no código e também farei comentários sobre suas percepções sobre o assunto. Estarei aberto a feedbacks e discussões sobre o tema.

**Peço que mantenha e não altere os comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender o meu feedback no seu notebook. Funciona assim:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Alerta não crítico, mas que pode ser corrigido para melhoria geral no seu código/análise.
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Erro que precisa ser arrumado, caso contrário seu projeto **não** será aceito.
</div>

Você pode interagir comigo através dessa célula:
<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-danger">
<b> Comentário geral do revisor v1</b> <a class="tocSkip"></a>
    
Obrigado por enviar o seu projeto e pelo esforço de chegar até aqui. O seu projeto possui alguns pontos bem interessantes, dos quais eu destaco:
    
- Código bem simples e estruturado 
- Comentários sucintos e bem objetivos
    
Entretanto, deixei alguns comentários em alguns trechos que precisam de ajuste. Peço que trabalhe nesses pontos para avançarmos.
    
Qualquer dúvida, pode contar comigo.   
    
    
**Até breve!**

</div>

<div class="alert alert-block alert-info">
<b>Obrigado pela revisão Luiz. Só para esclarecer um pouco sobre os ajustes, a parte onde precisa ter pelo menos 3 hiperparametros e 75% de acuracia não está na descrição do projeto, acredito que quando os outros alunos chegarem aqui todos irao mandar sem estes detalhes também. Concordo que sempre devemos melhorar nossos projetos, apenas acho que exigencias específicas precisam estar determinadas. É como se o cliente pedisse para comprar maçã e depois que nós entregassemos a sacola ele falasse que queria exatamente 7 maçãs. </b> <a class="tocSkip"></a>
</div>


<div class="alert alert-block alert-info">
<b> 
    O que nós temos na descrição é isto:
    
Investigue a qualidade de diferentes modelos alterando hiperparâmetros. 
    
Verifique a qualidade do modelo usando o conjunto de teste..
    </b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b> Comentário geral do revisor v2</b> <a class="tocSkip"></a>

Obrigado por enviar o seu projeto e fazer os ajustes sugeridos. Essa versão do seu trabalho ficou muito melhor! 

<br>

Sobre o score de `75%` de acurácia, não sabia que essa informação não estava disponível no enunciado do projeto, pois ela é um requisito na nossa rúbrica de avaliação. Porém, concordo com você que ficaria mais claro se essa informação estivesse disponível para vocês. Vou fazer essa sugestão internamente, obrigado!    
    
    
<br>
Te desejo uma jornada de muito sucesso e aprendizado.
    
<br>   
    
Qualquer dúvida, pode contar comigo.   
    
<br>  
    
**Até breve!**

</div>

# Modelos de machine learning testados no dataset users_behavior

In [20]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV

# Carregando e lendo os dados

In [2]:
df = pd.read_csv('/datasets/users_behavior.csv')

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

- O conjunto de dados foi carregado corretamente
- Os pacotes foram importados corretamente em células separadas
</div>

In [3]:
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [4]:
df.duplicated().sum()

0

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [6]:
df.isna().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

In [7]:
df[df['is_ultra'] == 1].describe().reset_index()

index       calls      minutes    messages       mb_used  is_ultra
0  count  985.000000   985.000000  985.000000    985.000000     985.0
1   mean   73.392893   511.224569   49.363452  19468.823228       1.0
2    std   43.916853   308.031100   47.804457  10087.178654       0.0
3    min    0.000000     0.000000    0.000000      0.000000       1.0
4    25%   41.000000   276.030000    6.000000  11770.280000       1.0
5    50%   74.000000   502.550000   38.000000  19308.010000       1.0
6    75%  104.000000   730.050000   79.000000  26837.720000       1.0
7    max  244.000000  1632.060000  224.000000  49745.730000       1.0

In [8]:
df[df['is_ultra'] == 0].describe().reset_index()

index        calls      minutes     messages       mb_used  is_ultra
0  count  2229.000000  2229.000000  2229.000000   2229.000000    2229.0
1   mean    58.463437   405.942952    33.384029  16208.466949       0.0
2    std    25.939858   184.512604    28.227876   5870.498853       0.0
3    min     0.000000     0.000000     0.000000      0.000000       0.0
4    25%    40.000000   274.230000    10.000000  12643.050000       0.0
5    50%    60.000000   410.560000    28.000000  16506.930000       0.0
6    75%    76.000000   529.510000    51.000000  20043.060000       0.0
7    max   198.000000  1390.220000   143.000000  38552.620000       0.0

In [9]:
df[df['is_ultra']== 1].count()

calls       985
minutes     985
messages    985
mb_used     985
is_ultra    985
dtype: int64

In [10]:
df[df['is_ultra']== 0].count()

calls       2229
minutes     2229
messages    2229
mb_used     2229
is_ultra    2229
dtype: int64

In [11]:
df['is_ultra'].value_counts(normalize=True)

0    0.693528
1    0.306472
Name: is_ultra, dtype: float64

**Temos 985 planos ultra e 2229 planos smart na proporção de 31% para 69% respectivamente**

**Analisando os dados vemos que eles estão tratados ou então foram bem armazenados, sem valores ausentes, nulos ou erros de classe.**

**Usuários do plano ultra costumam usar um numero bem maior de dados do que os do plano smart, uma informação na qual podemos explorar para achar potenciais clientes para o plano ultra**

<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>
<s>
    
- Aqui você precisa fazer um estudo de valores ausentes também: `data.isna().sum())`  
- Você precisa fazer um estudo de balanceamento dos dados também. Qual a proporção de exemplos na classe `is_ultra==1`? Essa propriedade é importante em modelagem de aprendizado de máquina para embasar a escolha de modelos e técnicas de transformação dos dados. Sugestão:
    
```python
data['is_ultra'].value_counts(normalize=True)
...
```
- É necessário adicionar suas conclusões também sobre essa etapa
</s>
</div>

# Separando os dados em conjuntos de treino,validação e teste

In [12]:
features = df.drop('is_ultra', axis= 1)
target = df['is_ultra']

x_train, x_temp, y_train, y_temp = train_test_split(features, target, test_size=0.4, random_state=42)
x_val, x_test, y_val, y_test = train_test_split(x_temp, y_temp, test_size=0.5, random_state=42)

print('Quantidade no conjunto de treino:', x_train.shape)
print('Quantidade no conjunto de validação:', x_val.shape)
print('Quantidade no conjunto de teste:', x_test.shape)


Quantidade no conjunto de treino: (1928, 4)
Quantidade no conjunto de validação: (643, 4)
Quantidade no conjunto de teste: (643, 4)


**conjunto de dados divididos em 60% treinamento , 20% validação e 20% teste**

<div class="alert alert-block alert-warning">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

- Os dados foram separados em `treino`, `validação` e `teste` corretamente. Como sugestão, você poderia imprimir a quantidade de exemplos em cada conjunto.
</div>

# Qualidade dos modelos: qual faz a melhor previsão?

# DecisionTreeClassifier

In [13]:
best_depth = None
best_accuracy = 0.0

for depth in range(1, 16):
    for min_samples_leaf in range(1, 6):  
        model = DecisionTreeClassifier(random_state=42, 
                                       max_depth=depth, 
                                       min_samples_split=2, 
                                       min_samples_leaf=min_samples_leaf)
        model.fit(x_train, y_train)
        
        accuracy_train = model.score(x_train, y_train)
        accuracy_val = model.score(x_val, y_val)
        
        if accuracy_val > best_accuracy:
            best_accuracy = accuracy_val
            best_depth = depth
            best_min_samples_leaf = min_samples_leaf

print(f'Best Accuracy: {best_accuracy:.4f}')
print(f'Best Depth: {best_depth}')
print(f'Best Min Samples Leaf: {best_min_samples_leaf}')

Best Accuracy: 0.8009
Best Depth: 10
Best Min Samples Leaf: 2


**achamos neste modelo qual é a melhor profundidade para encontrar a melhor precisão com um hiperparametro de max_depth = 10 e com um numero minimo de 2 amostras por folha (este hiperparametro ajuda a ajustar o modelo para nao ficar nem sobreajustado e nem subajustado)**

# RandomForestClassifier

In [15]:
best_score =0
best_est =0

for est in range(1, 31):
    for depth in range(1, 11): 
        model = RandomForestClassifier(random_state=42, n_estimators=est, max_depth=depth)
        model.fit(x_train, y_train)
        score = model.score(x_val, y_val)
        
        if score > best_score:
            best_score = score
            best_est = est
            best_depth = depth
        
        
print('best_score: ', best_score, 'best_est:', best_est, 'best_depth:', best_depth)

best_score:  0.8195956454121306 best_est: 24 best_depth: 9


**testamos o numero de estimadores de 1 a 30 e max_depth entre 1 e 10 para achar qual é o numero com melhor precisão no modelo sendo n_estiamtor = 24 e max_depth = 9 o melhor hiperparametro para este modelo**

# RandomForestClassifier usado no conjunto de teste

In [17]:
model = RandomForestClassifier(random_state=42, n_estimators=24, max_depth=9)
model.fit(x_train, y_train)
score_test = model.score(x_test, y_test)
print(score_test)

0.8164852255054432


**Até então temos o randomforest como o modelo com melhor precisão de acertos**

# LogisticRegression

In [35]:

scale = StandardScaler()
x_train_scale = scale.fit_transform(x_train)
x_val_scale = scale.transform(x_val) 


model = LogisticRegression(random_state=42)

param_grid = {
    'solver': ['liblinear', 'saga'], 
    'C': [0.01, 0.1, 1, 10, 100],
    'penalty': ['l1', 'l2']  
}

grid_search = GridSearchCV(estimator=model,  param_grid=param_grid,  cv=5, scoring='accuracy', n_jobs=-1)  

grid_search.fit(x_train_scale, y_train)

# Melhor modelo selecionado pelo GridSearchCV
best_model = grid_search.best_estimator_

score_train = best_model.score(x_train_scale, y_train)
score_valid = best_model.score(x_val_scale, y_val)

print('Precisão de treino:', score_train)
print('Precisão de validação:', score_valid)
print('Melhores parâmetros:', grid_search.best_params_)

Precisão de treino: 0.7422199170124482
Precisão de validação: 0.7387247278382582
Melhores parâmetros: {'C': 0.1, 'penalty': 'l2', 'solver': 'saga'}


**Usamos a regressão logistica por estamos tentando achar uma variavel categorica, caso fossem variaveis numericas continuas usariamos a regressão liner**

**Apesar de testarmos varios hiperparametros adicionais para este modelo ele ficou com a precisão abaixo do nossos outros dois modelos**

<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>
<s>
    
O modelo `LogisticRegression()` é baseado em pesos, ou seja, se você tiver dados em escalas muito diferentes, essa característica irá impactar no modelo treinado. Por esse motivo, é preciso normalizar os dados para esse modelo especificamente dos modelos que você utilizou (você pode usar `StandardScaler()`, por exemplo).
</s>
</div>

<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>
<s>
    
Aqui nesta seção, precisamos de alguns ajustes:
    
- [x] Pelo menos 2 modelos foram testados
- [x] Pelo menos 3 hiperparâmetros foram estudados
- [x] A etapa de estudo de hiperparâmetros precisa ser feita no conjunto de **validação** e não de teste. O conjunto de teste deve ser usado apenas para validação final dos modelos
- [x] A acurácia de pelo menos `75%` precisa ser atingida no conjunto **de teste**
- Dica: Você pode usar métodos como `Randomsearch` para fazer o estudo de hiperparâmetros de forma automatizada. Exemplo:
    ```python
        param_grids = {
            'DecisionTree': {'max_depth': [None, 5, 10, 15], 'min_samples_split': [2, 5, 10]},
            'RandomForest': {'n_estimators': [10, 50, 100], 'max_depth': [None, 5, 10, 15]},
            'KNeighbors': {'n_neighbors': [3, 5, 7], 'weights': ['uniform', 'distance']}
        }
    ...

        grid_search = GridSearchCV(model, param_grids[model_name], cv=5, scoring='accuracy')
        grid_search.fit(features_train, target_train)
        ...
    ```
    
**Obs.**: a prova real deve ser fazer o teste do melhor modelo no conjunto de *teste**. Na etapa de estudo dos hiperparâmetros, você deve usar o conjunto de validação e só então, por fim, você deve usar o conjunto de dados de teste.
</s>
</div>

# Prova real do melhor modelo: RandomForest

In [36]:
best_score =0
best_est =0

for est in range(1, 31):
    for depth in range(1, 11): 
        model = RandomForestClassifier(random_state=42, n_estimators=est, max_depth=depth)
        model.fit(x_train, y_train)
        score = model.score(x_test, y_test)
        
        if score > best_score:
            best_score = score
            best_est = est
            best_depth = depth
        
        
print('best_score: ', best_score, 'best_est:', best_est, 'best_depth:', best_depth)

best_score:  0.8304821150855366 best_est: 10 best_depth: 6


**Aqui na prova real fizemos o teste do nosso modelo mais preciso com o conjunto de teste e encontramos uma acurácia ainda maior do que no conjunto de treinamento com n_estimators = 10 e max_depth = 6, confirmando que este é o modelo mais preciso.**

# Conclusão

**Se quisermos o modelo com melhor precisão de previsão e tivermos capacidade computacional para suportal tal feito usaremos o modelo de RandomForest.**

**Caso não tenhamos tal poder, usaremos o DecisionTreeClassifier com o hiperparametro max_depth = 10 e min_samples_leaf = 2**